libherokubuildpack/src/output/util.rs

```diff
@@ -0,0 +1,88 @@
+use lazy_static::lazy_static;
+lazy_static! {
+    static ref TRAILING_WHITESPACE_RE: regex::Regex = regex::Regex::new(r"\s+$").expect("clippy");
+}
+
+/// Iterator yielding every line in a string. The line includes newline character(s).
+///
+/// <https://stackoverflow.com/a/40457615>
+///
+/// The problem this solves is when iterating over lines of a string, the whitespace may be significant.
+/// For example if you want to split a string and then get the original string back then calling
+/// `lines().collect<Vec<_>>().join("\n")` will never preserve trailing newlines.
+///
+/// There's another option to `lines().fold(String::new(), |s, l| s + l + "\n")`, but that
+/// always adds a trailing newline even if the original string doesn't have one.
+pub(crate) struct LinesWithEndings<'a> {
+    input: &'a str,
+}
+
+impl<'a> LinesWithEndings<'a> {
+    pub(crate) fn from(input: &'a str) -> LinesWithEndings<'a> {
+        LinesWithEndings { input }
+    }
+}
+
+impl<'a> Iterator for LinesWithEndings<'a> {
+    type Item = &'a str;
+
+    #[inline]
+    fn next(&mut self) -> Option<&'a str> {
+        if self.input.is_empty() {
+            return None;
+        }
+        let split = self.input.find('\n').map_or(self.input.len(), |i| i + 1);
+
+        let (line, rest) = self.input.split_at(split);
+        self.input = rest;
+        Some(line)
+    }
+}
+
+/// Removes trailing whitespace from lines
+///
+/// Useful because most editors strip trailing whitespace (in test fixtures)
+/// but commands <https://github.com/heroku/libcnb.rs/issues/582> emit newlines
+/// with leading spaces. These can be sanatized by removing trailing whitespace.
+#[allow(dead_code)]
+pub(crate) fn strip_trailing_whitespace(s: impl AsRef<str>) -> String {
```

Comment:
I'm not sure I really understand the need for this feature?

It also uses the regex crate, which isn't ideal.

At the very least we should switch this to a non-regex solution, or else remove it outright.

Reply:
This should be functionally equivalent to [trim_end](https://doc.rust-lang.org/std/primitive.str.html#method.trim_end)

Reply:
This function operates on all lines within the string, not just the trailing whitespace on the string itself. It is only used for testing. I'm:

- [ ] Changing the name to include "lines" and to reflect the similarity to `trim_end` I'll call it `trim_end_lines`
- [ ] Make it `cfg(test)` instead of `allow(dead_code)`

Reply:
- [ ] Remove regex and lazy_static crates